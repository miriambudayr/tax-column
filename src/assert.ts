export function assert(v: any, why = ""): asserts v {
  if (!v) {
    const error = new Error(`Assertion Failed: ${why}`);
    error.name = "AssertionFailure";

    console.error("AssertionFailure", error);
    throw error;
  }
}
