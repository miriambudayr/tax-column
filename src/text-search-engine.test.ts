import { DB_PATH, TextSearchEngine } from ".";
import dotenv from "dotenv";
import fs from "fs";
import { assert } from "./assert";
dotenv.config();

const TEST_DATA_DIR = "src/test_data";

describe("Text Search Engine", () => {
  describe("Database Setup", () => {
    beforeEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    afterEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    it("should create the SQLite database file", () => {
      new TextSearchEngine();
      const dbPath = DB_PATH;
      expect(fs.existsSync(dbPath)).toBe(true);
    });
  });

  describe("File Indexing and Search", () => {
    beforeEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    afterEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    it("should index and search a file", () => {
      const fileName = `${TEST_DATA_DIR}/lorem-ipsum.txt`;
      const searchEngine = new TextSearchEngine();
      searchEngine.indexFile(fileName);

      const files = searchEngine.search("metus");

      expect(files.length).toBeGreaterThan(0);

      if (files) {
        for (const file of files) {
          expect(file).toEqual(fileName);
        }
      }
    });

    it("should search a phrase", () => {
      const fileName = `${TEST_DATA_DIR}/lorem-ipsum.txt`;
      const searchEngine = new TextSearchEngine();
      searchEngine.indexFile(fileName);

      const files = searchEngine.searchPhrase("euismod ornare ultrices");

      expect(files.length).toBeGreaterThan(0);

      if (files) {
        for (const file of files) {
          expect(file).toEqual(fileName);
        }
      }
    });

    it("should tokenize a directory", () => {
      // TODO: try using actual data sample once this works. The dev feedback loop was way too slow
      // to use the sample data for now.
      // const dirName = process.env.SAMPLE_DATA_DIR;
      // assert(dirName, "SAMPLE_DATA_DIR environment variable not found");

      const dirName = TEST_DATA_DIR;
      const searchEngine = new TextSearchEngine();
      searchEngine.recursivelyIndexDirectory(dirName);

      let files = searchEngine.search("porta");
      expect(Array.from(files!)).toMatchSnapshot();

      files = searchEngine.searchPhrase("porta magna urna et elit");
      expect(Array.from(files!)).toMatchSnapshot();
    });

    it("should recursively tokenize a directory", () => {
      const dirName = TEST_DATA_DIR;
      const searchEngine = new TextSearchEngine();
      searchEngine.recursivelyIndexDirectory(dirName);

      let files = searchEngine.search("simmons");
      expect(Array.from(files!)).toMatchSnapshot();

      files = searchEngine.searchPhrase("dinner hosted by matt simmons");
      expect(Array.from(files!)).toMatchSnapshot();
    });

    it("should use case-insensitive search", () => {
      const dirName = TEST_DATA_DIR;
      const searchEngine = new TextSearchEngine();
      searchEngine.recursivelyIndexDirectory(dirName);

      let files = searchEngine.search("Simmons");
      expect(Array.from(files!)).toMatchSnapshot();

      files = searchEngine.searchPhrase("Dinner hosted by Matt Simmons");
      expect(Array.from(files!)).toMatchSnapshot();
    });
  });

  describe("Prefix Search (Search-as-You-Type)", () => {
    beforeEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    afterEach(() => {
      if (fs.existsSync(DB_PATH)) {
        fs.unlinkSync(DB_PATH);
      }
    });

    it("should return files and words containing a prefix", () => {
      const dirName = TEST_DATA_DIR;
      const searchEngine = new TextSearchEngine();
      searchEngine.recursivelyIndexDirectory(dirName);

      const results = searchEngine.searchPrefix("pres");

      expect(results.length).toBeGreaterThan(0);
      expect(results).toMatchSnapshot();
    });
  });
});
