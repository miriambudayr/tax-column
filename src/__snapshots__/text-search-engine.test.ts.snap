// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Search Engine File Indexing and Search should recursively tokenize a directory 1`] = `
[
  "src/test_data/calendar/1.",
  "src/test_data/calendar/1.",
]
`;

exports[`Text Search Engine File Indexing and Search should recursively tokenize a directory 2`] = `
[
  "src/test_data/calendar/1.",
]
`;

exports[`Text Search Engine File Indexing and Search should tokenize a directory 1`] = `
[
  "src/test_data/lorem-ipsum-2.txt",
  "src/test_data/lorem-ipsum.txt",
  "src/test_data/lorem-ipsum.txt",
  "src/test_data/lorem-ipsum.txt",
  "src/test_data/lorem-ipsum.txt",
  "src/test_data/lorem-ipsum.txt",
]
`;

exports[`Text Search Engine File Indexing and Search should tokenize a directory 2`] = `
[
  "src/test_data/lorem-ipsum.txt",
]
`;

exports[`Text Search Engine File Indexing and Search should use case-insensitive search 1`] = `
[
  "src/test_data/calendar/1.",
  "src/test_data/calendar/1.",
]
`;

exports[`Text Search Engine File Indexing and Search should use case-insensitive search 2`] = `
[
  "src/test_data/calendar/1.",
]
`;

exports[`Text Search Engine Prefix Search (Search-as-You-Type) should return files and words containing a prefix 1`] = `
[
  {
    "file": "src/test_data/calendar/1.",
    "matchedWord": "president",
    "position": undefined,
  },
  {
    "file": "src/test_data/calendar/2.",
    "matchedWord": "president",
    "position": undefined,
  },
  {
    "file": "src/test_data/calendar/2.",
    "matchedWord": "presentation",
    "position": undefined,
  },
]
`;
